File: Java/ListaEncadeada.java
/*  Lista encadeada em Java como uma subclasse de Lista */
/* Usei como base de polimorfismo em java o seguinte link: https://www.devmedia.com.br/uso-de-polimorfismo-em-java/26140 */

public class ListaEncadeada extends Lista{
    private Node inicio;
    public ListaEncadeada(){
        this.inicio = null;
    }

    @Override
    public void insere(int value){
        Node newNode = new Node(value);
        newNode.setNext(this.inicio);
        this.inicio = newNode;
        this.setSize(this.getSize() + 1);
    }

    @Override
    public void retira(int value) {
        if (this.inicio == null) {
            return;
        }
        Node nodeAtual = this.inicio;
        Node nodeAnterior = null;
        int size = this.getSize();
        while(nodeAtual != null){
            if(nodeAtual.getValue() == value){
                if(nodeAnterior == null){ //caso especial se estiver no inicio da lista
                    this.inicio.setNext(nodeAtual.getNext());
                } else{
                    nodeAnterior.setNext(nodeAtual.getNext());
                }
                this.setSize(size--);
                return;
            }
            nodeAnterior = nodeAtual;
            nodeAtual = nodeAnterior.getNext();
        }
    }

    @Override
    public int primeiro(){
        if(this.getSize() == 0){
            return -1;   
        }
        return this.inicio.getValue();
    }

    @Override
    public int ultimo(){
        if(this.getSize() == 0){
            return -1;   
        }
        Node noAtual = this.inicio;
        while (noAtual.getNext() != null) {
            noAtual = noAtual.getNext();
        }

        return noAtual.getValue();
    }

 
}

File: C++/ListaEncadeada.cpp
#include "ListaEncadeada.h"

ListaEncadeada::ListaEncadeada() : inicio(nullptr) {}

ListaEncadeada::~ListaEncadeada() {
    Node* current = inicio;
    while (current != nullptr) {
        Node* next = current->next;
        delete current;
        current = next;
    }
}

void ListaEncadeada::insere(int value) {
    Node* newNode = new Node(value);
    newNode->next = inicio;
    inicio = newNode;
    set_size(get_size() + 1);
}

void ListaEncadeada::retira(int value) {
    if (inicio == nullptr) {
        return;
    }

    Node* current = inicio;
    Node* previous = nullptr;

    while (current != nullptr) {
        if (current->value == value) {
            if (previous == nullptr) {
                inicio = current->next;
            } else {
                previous->next = current->next;
            }
            set_size(get_size() - 1);
            delete current;
            return;
        }
        previous = current;
        current = previous->next;
    }
}

int ListaEncadeada::primeiro() {
    if (get_size() == 0 || inicio == nullptr) {
        return -1;
    }
    return inicio->value;
}

int ListaEncadeada::ultimo() {
    if (get_size() == 0 || inicio == nullptr) {
        return -1;
    }

    Node* current = inicio;
    while (current->next != nullptr) {
        current = current->next;
    }

    return current->value;
}
